@PaulXiCao This seems to be failing with many error messages that seem legitimate: 

```
__math/hypot.h:74:15: error: magnitude of floating-point constant too large for type 'long double'; maximum is 1.7976931348623157E+308 [-Werror,-Wliteral-range]
      return {0x1.0p+8'190l, 0x1.0p-9'000l};
              ^
```

--- a/libcxx/include/__math/hypot.h
+++ b/libcxx/include/__math/hypot.h
@@ -66,13 +66,18 @@ _LIBCPP_HIDE_FROM_ABI std::pair<_Real, _Real> __hypot_factors() {
     return {0x1.0p+510, 0x1.0p-600};
   } else { // long double
     static_assert(std::is_same_v<_Real, long double>);
-    if constexpr (sizeof(_Real) == sizeof(double))
-      return static_cast<std::pair<_Real, _Real>>(__math::__hypot_factors<double>());
-    else {
-      static_assert(-16'381 == std::numeric_limits<_Real>::min_exponent);
-      static_assert(+16'384 == std::numeric_limits<_Real>::max_exponent);
-      return {0x1.0p+8'190l, 0x1.0p-9'000l};
-    }
+
+    // preprocessor guard necessary, otherwise literals (e.g. `0x1.0p+8'190l`) throw warnings even when shielded by `if
+    // constexpr`
+#  if __DBL_MAX_EXP__ == __LDBL_MAX_EXP__
+    static_assert(sizeof(_Real) == sizeof(double));
+    return static_cast<std::pair<_Real, _Real>>(__math::__hypot_factors<double>());
+#  else
+    static_assert(sizeof(_Real) > sizeof(double));
+    static_assert(-16'381 == std::numeric_limits<_Real>::min_exponent);
+    static_assert(+16'384 == std::numeric_limits<_Real>::max_exponent);
+    return {0x1.0p+8'190l, 0x1.0p-9'000l};
+#  endif
   }
 }
 

--- a/libcxx/include/__math/hypot.h
+++ b/libcxx/include/__math/hypot.h
@@ -87,7 +87,7 @@ _LIBCPP_HIDE_FROM_ABI std::pair<_Real, _Real> __hypot_factors() {
 // See https://github.com/llvm/llvm-project/issues/92782 for a detailed discussion and summary.
 template <class _Real>
 _LIBCPP_HIDE_FROM_ABI _Real __hypot(_Real __x, _Real __y, _Real __z) {
-  const _Real __max_abs = std::max({__math::fabs(__x), __math::fabs(__y), __math::fabs(__z)});
+  const _Real __max_abs = std::max(__math::fabs(__x), std::max(__math::fabs(__y), __math::fabs(__z)));
   const auto [__overflow_threshold, __overflow_scale] = __math::__hypot_factors<_Real>();
   _Real __scale;
   if (__max_abs > __overflow_threshold) { // x*x + y*y + z*z might overflow

--- a/libcxx/test/std/numerics/c.math/cmath.pass.cpp
+++ b/libcxx/test/std/numerics/c.math/cmath.pass.cpp
@@ -1117,41 +1117,41 @@ void test_fmin()
 }
 
 struct TestHypot3 {
-    template <class Real>
-    void operator()() const {
-        const auto check = [](Real elem, Real abs_tol) {
-            assert(std::isfinite(std::hypot(elem, Real(0), Real(0))));
-            assert(fptest_close(std::hypot(elem, Real(0), Real(0)), elem, abs_tol));
-            assert(std::isfinite(std::hypot(elem, elem, Real(0))));
-            assert(fptest_close(std::hypot(elem, elem, Real(0)), std::sqrt(Real(2)) * elem, abs_tol));
-            assert(std::isfinite(std::hypot(elem, elem, elem)));
-            assert(fptest_close(std::hypot(elem, elem, elem), std::sqrt(Real(3)) * elem, abs_tol));
-        };
+  template <class Real>
+  void operator()() const {
+    const auto check = [](Real elem, Real abs_tol) {
+      assert(std::isfinite(std::hypot(elem, Real(0), Real(0))));
+      assert(fptest_close(std::hypot(elem, Real(0), Real(0)), elem, abs_tol));
+      assert(std::isfinite(std::hypot(elem, elem, Real(0))));
+      assert(fptest_close(std::hypot(elem, elem, Real(0)), std::sqrt(Real(2)) * elem, abs_tol));
+      assert(std::isfinite(std::hypot(elem, elem, elem)));
+      assert(fptest_close(std::hypot(elem, elem, elem), std::sqrt(Real(3)) * elem, abs_tol));
+    };
 
-        { // check for overflow
-            const auto [elem, abs_tol] = []() -> std::array<Real, 2> {
-                if constexpr (std::is_same_v<Real, float>)
-                    return {1e20f, 1e16f};
-                else if constexpr (std::is_same_v<Real, double>)
-                    return {1e300, 1e287};
-                else // long double
-                    return {1e4000l, 1e3985l};
-            }();
-            check(elem, abs_tol);
-        }
-
-        { // check for underflow
-            const auto [elem, abs_tol] = []() -> std::array<Real, 2> {
-                if constexpr (std::is_same_v<Real, float>)
-                    return {1e-20f, 1e-24f};
-                else if constexpr (std::is_same_v<Real, double>)
-                    return {1e-287, 1e-300};
-                else // long double
-                    return {1e-3985l, 1e-4000l};
-            }();
-            check(elem, abs_tol);
-        }
+    { // check for overflow
+      const auto [elem, abs_tol] = []() -> std::array<Real, 2> {
+        if constexpr (std::is_same_v<Real, float>)
+          return {1e20f, 1e16f};
+        else if constexpr (std::is_same_v<Real, double>)
+          return {1e300, 1e287};
+        else // long double
+          return {1e4000l, 1e3985l};
+      }();
+      check(elem, abs_tol);
     }
+
+    { // check for underflow
+      const auto [elem, abs_tol] = []() -> std::array<Real, 2> {
+        if constexpr (std::is_same_v<Real, float>)
+          return {1e-20f, 1e-24f};
+        else if constexpr (std::is_same_v<Real, double>)
+          return {1e-287, 1e-300};
+        else // long double
+          return {1e-3985l, 1e-4000l};
+      }();
+      check(elem, abs_tol);
+    }
+  }
 };
 
 void test_hypot()

